### ISSUE TITLE
MariaDB: long CHECK constraint clauses truncated in information_schema

### ISSUE THREAD

**Issue by elvenarg** [skeema/skeema#165]
The issue happens if the string length inside the CHECK() is too long, this can be accomplished with a long column name AND/OR multiple clauses inside the check.
This happens with both, ALTER and CREATE.

**Environment**
- Skeema version: skeema version 1.5.1-community, commit 365b56853672152055d75909ee63bfc35f286c28, released 2021-04-30T02:21:04Z
- MariaDB 10.2 Linux

**To Reproduce**
Have two databases, on the main one add the following table and update schema locally:

create table checktest2(a_very_long_column_name_to_make_it_cut_the_string int unsigned not null check (`a_very_long_column_name_to_make_it_cut_the_string` > 19 or `a_very_long_column_name_to_make_it_cut_the_string` = 1));

Run schema diff and we receive:

2021-05-31 14:10:39 [WARN]  Skipping table `checktest2`: unable to generate DDL due to use of unsupported features. Use --debug for more information.
2021-05-31 14:10:39 [DEBUG] --- Expected CREATE
2021-05-31 14:10:39 [DEBUG] +++ MySQL-actual SHOW CREATE
2021-05-31 14:10:39 [DEBUG] @@ -2 +2 @@
2021-05-31 14:10:39 [DEBUG] -  `a_very_long_column_name_to_make_it_cut_the_string` int(10) unsigned NOT NULL CHECK (`a_very_long_column_name_to_make_it_cut_the_string` > 19 or `a_v)
2021-05-31 14:10:39 [DEBUG] +  `a_very_long_column_name_to_make_it_cut_the_string` int(10) unsigned NOT NULL CHECK (`a_very_long_column_name_to_make_it_cut_the_string` > 19 or
                            `a_very_long_column_name_to_make_it_cut_the_string` = 1)

**Expected behavior**
The correct create/alter table

Thanks.

**Comment by evanelias** [comment]
Thank you for the issue report. The root cause here is actually a MariaDB bug, rather than a Skeema bug. It appears that MariaDB only allocates a max of 64 characters to the `information_schema.CHECK_CONSTRAINTS.CHECK_CLAUSE` column:

```
MariaDB [(none)]> desc information_schema.check_constraints;
+--------------------+--------------+------+-----+---------+-------+
| Field              | Type         | Null | Key | Default | Extra |
+--------------------+--------------+------+-----+---------+-------+
| CONSTRAINT_CATALOG | varchar(512) | NO   |     |         |       |
| CONSTRAINT_SCHEMA  | varchar(64)  | NO   |     |         |       |
| TABLE_NAME         | varchar(64)  | NO   |     |         |       |
| CONSTRAINT_NAME    | varchar(64)  | NO   |     |         |       |
| CHECK_CLAUSE       | varchar(64)  | NO   |     |         |       |
+--------------------+--------------+------+-----+---------+-------+
5 rows in set (0.01 sec)
```

Unfortunately this means that any tool relying on `information_schema` (including Skeema) is going to see truncated check constraint clauses if the clause is over 64 chars in MariaDB. fwiw the same issue does not happen in MySQL, since MySQL uses `longtext` for this information_schema column.

It should be possible to implement a workaround in Skeema to parse long check constraint clauses from `SHOW CREATE TABLE` in MariaDB, rather than relying on `information_schema.check_constraints` as-is. However this rapidly gets messy and complicated when combined with the other existing workarounds to deal with check constraint edge cases, such as how MariaDB exposes inline column checks in a weird way in information_schema. So I can't predict an ETA for when a Skeema workaround may be implemented.

> This happens with both, ALTER and CREATE.

Are you sure about this? I would be extremely surprised if `CREATE` is affected, since this is a very different code path which does not use `information_schema`. In other words, if a table with a long check constraint exists in the filesystem but does not exist in the target DB yet, I would expect that `skeema diff` and `skeema push` still correctly emit the right `CREATE TABLE`.

The message `unable to generate DDL due to use of unsupported features` only ever happens with generation of `ALTER`; in this situation you can always still [create the table if it does not already exist](https://www.skeema.io/docs/requirements/#unsupported-for-alter-table) even if it uses unsupported features.

**Comment by elvenarg** [comment]
Hi Evan, thanks for the detailed and crystal clear explanation of the issue, never thought on checking the "check_constraints" table.

About CREATE/ALTER you are right, the issue is only with ALTER, just checked it.

**Comment by evanelias** [comment]
I've implemented a workaround for this by re-parsing the full check clause from SHOW CREATE TABLE when necessary. I'll push the commit momentarily. The next binary release might not be for a month or two though.

In the process of testing my patch, I also noticed that MariaDB actually [fixed the root problem](https://jira.mariadb.org/browse/MDEV-24139) in their February releases: MariaDB 10.2.37, 10.3.28, 10.4.18, and 10.5.9. So upgrading to a newer MariaDB patch release is another possible solution for anyone hitting this.
